src/ray/core_worker/task_execution.cc

```diff
@@ -69,10 +59,13 @@ Status CoreWorkerTaskExecutionInterface::ExecuteTask(
 
 void CoreWorkerTaskExecutionInterface::Run() {
   // Run main IO service.
-  main_service_.run();
+  main_service_->run();
+}
 
-  // should never reach here.
-  RAY_LOG(FATAL) << "should never reach here after running main io service";
+void CoreWorkerTaskExecutionInterface::Stop() {
+  // Stop main IO service.
+  std::shared_ptr<boost::asio::io_service> main_service = main_service_;
+  main_service_->post([main_service]() { main_service->stop(); });
```

Comment:
Comment added and also post it here.
```
// Delay the execution of io_service::stop() to avoid deadlock if
// CoreWorkerTaskExecutionInterface::Stop is called inside a task.
```